dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,136 @@
+                                  ]
+
+parse :: Text -> M.Map FilePos Tag
+parse t = M.fromAscList . map (\(FP ln c, Tag _ term) ->
+              let (lsl, l) = mls M.! ln in
+              (FP ln (lsl + c), Tag (T.take c l) term)) . M.assocs $
+          parseExpr e (FP 0 0) M.empty
+    where e = case exprFromText "" t of
+                  Left _ -> None
+                  Right r -> r
+          mls = M.fromList . fst . foldl (\(ls, lsl) (n, l) ->
+                                             let lsl' = lsl + 1 + T.length l in
+                                             ((n, (lsl, l)):ls, lsl'))
+                                        ([], 0) . zip [1..] $ T.lines t
+
+parseExpr :: Expr Src a -> FilePos -> M.Map FilePos Tag -> M.Map FilePos Tag
+parseExpr (Let b e) lpos  mts = parseExpr e lpos (parseBinding b mts)
+-- | > Annot x t                                ~  x : t
+parseExpr (Annot e1 e2) lpos mts = parseExpr e1 lpos (parseExpr e2 lpos mts)
+-- | > Record       [(k1, t1), (k2, t2)]        ~  { k1 : t1, k2 : t1 }
+parseExpr (Record mr) lpos mts = foldr (\k mts' ->
+    M.insert lpos (Tag "" k) mts') mts . DM.keys $  mr
+-- | > RecordLit    [(k1, v1), (k2, v2)]        ~  { k1 = v1, k2 = v2 }
+parseExpr (RecordLit mr) lpos mts = foldr (\k mts' ->
+    M.insert lpos (Tag "" k) mts') mts . DM.keys $ mr
+-- | > Union        [(k1, Just t1), (k2, Nothing)] ~  < k1 : t1 | k2 >
+parseExpr (Union mmr) lpos mts = foldr (\k mts' ->
+    M.insert lpos (Tag "" k) mts') mts . DM.keys $ mmr
+-- | > Note s x                                 ~  e
+parseExpr (Note s e) _ mts = parseExpr e (srcToFilePos s) mts
+parseExpr _ _ mts = mts
+
+parseBinding :: Binding Src a -> M.Map FilePos Tag -> M.Map FilePos Tag
+parseBinding b = M.insert (maybeSrcToFilePos . bindingSrc0 $ b)
+                          (Tag "" var)
+    where var = variable b
+
+maybeSrcToFilePos :: Maybe Src -> FilePos
+maybeSrcToFilePos Nothing = FP 0 0
+maybeSrcToFilePos (Just s) = srcToFilePos s
+
+srcToFilePos :: Src -> FilePos
+srcToFilePos s = FP line column
+    where ssp = srcStart s
+          line = unPos . sourceLine $ ssp
+          column = unPos . sourceColumn $ ssp
+
+showTags :: Tags -> Text
+showTags (Tags ts) = T.concat . M.elems . M.mapWithKey showFileTags $ ts
+
+showFileTags :: FilePath -> M.Map FilePos Tag -> T.Text
+showFileTags f ts = "\x0c\n" <> T.pack f <> "," <> (showInt . T.length) cs <> "\n" <> cs
+    where cs = T.concat . M.elems . M.mapWithKey showPosTag $ ts
+
+showPosTag :: FilePos -> Tag -> Text
+showPosTag (FP line shift) (Tag def term) = def <>"\x7f" <> term <> "\x01" <> showInt line <>
+                                             "," <> showInt shift <> "\n"
+
+showInt :: Int -> Text
+showInt = T.pack . show
+
+{-| from https://github.com/MarcWeber/hasktags/blob/master/src/Hasktags.hs
+ -| suffixes: [".dhall"], use "" to match all files
+-}
+dirToFiles :: Bool -> [String] -> FilePath -> IO [ FilePath ]
+dirToFiles _ _ "STDIN" = lines <$> getContents
```

Comment:
Is this alternative being used?

Reply:
I changed how it works.
So if all parameters are omitted, `dhall tags` will read list of files from stdin, so it's possible to redirect output of find to input of `dhall tags` and tool will write tags into `tags` file by default. So for dada project user should type: `dhall tags --path . --suffixes ""` and for `dhall-concourse` (I just found it on github): `dhall tags --path ."
Then you are ready to use Ctrl+] in vi and M-. in emacs.